### ISSUE TITLE
Reproduce deepdanbooru_v3 in PyTorch

### ISSUE THREAD

**Issue by ydk-tellurion** [KichangKim/DeepDanbooru#43]
Hi, I'm deeply impressed by your project and want to devote your model to my PyTorch project.
But when I used PyTorch to reproduce your project, my model performs much worse than yours after training. It is only able to predict high-frequency tags such as "solo, 1girl, highres", missing tags such as "blue eyes, green hair".  It seems to caused by the long tailed distribution of danbooru dataset, but I cannot fix it.

Have you met similar problems before? Did you solve it with some improvements on loss function or some trianing tricks?

I'm using the model custom deepdanbooru_v3 and danbooru2020 dataset with 7000 one-hot labels.
optimizier, loss function and other settings are the same as your default ones.

Thank you!

**Comment by YHJ2c95** [comment]
@KichangKim  @ydk-tellurion

hi 

after [survey danbooru's tag](https://gist.github.com/YHJ2c95/b03658740d45da98830f77e2678cd93a)
I think multi-label classification not a good 

### tag self with semantic, but is for human, as dataset is images bucket/collection
Concepts that one cannot describe / not presented / too technical terminology, no (this serious, poorly trained models, few downstream task)
Or even, [nothing learned at all](https://twitter.com/pbaylies/status/1439283298451312641)
perhaps can add some pseudo could from unsupervised cluster Can improved

There are tags for non-original characters that are bias, i.e. character traits

There are very few valid tags, (so why limit dan to 2, there are many "images" and two tags are enough to search)

There is no information about the components of the tag, [for example](https://anilist.co/anime/12679/Joshiraku/#:~:text=%E3%82%87%E3%81%97%E3%82%89%E3%81%8F-,Tags,-Cute%20Girls%20Doing)

The tag set is not strongly delineated and there is repetition of meaning

New tags are difficult to synchronise with earlier images 

### web image classification
I think the danbooru tagging task is a web image classification, which is very different from imagenet/CAPTCHA.

One is that imagenet/CAPTCHA is very close to objects, whereas web images are not like that

Secondly, if you remove the tagging literally, then the whole dataset should be just a bucket/collection, with sets of different themes of different intensities

Extend this further as a group, which can be subsetted and merged

Apply contrast learning again, instead of using single image view's contrast, turn it into a group-to-group contrast

The inspiration is to give a bunch of images and then [guess the tag game](https://github.com/lynn/koakuma) and [paper](https://openreview.net/forum?id=1IiJQTDpuG)

**Comment by KichangKim** [comment]
Yes, simple multi-label classification has many limitations for semantic recognition. That is why I removed "copyright tags" from training data.
